tensorflow_addons/layers/spectral_normalization.py

```diff
@@ -0,0 +1,128 @@
+# Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# =============================================================================
+
+import tensorflow as tf
+from typeguard import typechecked
+
+
+@tf.keras.utils.register_keras_serializable(package="Addons")
+class SpectralNormalization(tf.keras.layers.Wrapper):
+    """This wrapper controls the Lipschitz constant of the layer by
+       constraining its spectral norm.
+    This stabilizes the training of GANs.
+    Spectral Normalization for Generative Adversarial Networks:
+    https://arxiv.org/abs/1802.05957
+    Takeru Miyato, Toshiki Kataoka, Masanori Koyama, Yuichi Yoshida (2018)
+    SpectralNormalization wrapper works for keras and tf layers.
+    ```python
+      net = SpectralNormalization(
+          tf.keras.layers.Conv2D(2, 2, activation="relu"),
+          input_shape=(32, 32, 3))(x)
+      net = SpectralNormalization(
+          tf.keras.layers.Conv2D(16, 5, activation="relu"))(net)
+      net = SpectralNormalization(
+          tf.keras.layers.Dense(120, activation="relu"))(net)
+      net = SpectralNormalization(
+          tf.keras.layers.Dense(n_classes))(net)
+    ```
+    Arguments:
+      layer: a layer instance.
+    Raises:
+      AssertionError: If not initialized with a `Layer` instance.
+      ValueError: If initialized with negative `power_iterations`
+      AttributeError: If `Layer` does not contain a `kernel` or `embeddings` of weights
+    """
+
+    @typechecked
+    def __init__(self, layer: tf.keras.layers, power_iterations: int = 1, **kwargs):
+        super(SpectralNormalization, self).__init__(layer, **kwargs)
+        if power_iterations <= 0:
+            raise ValueError(
+                "`power_iterations` should be greater than zero, got "
+                "`power_iterations={}`".format(power_iterations)
+            )
+        self.power_iterations = power_iterations
+        self._initialized = False
+        self._track_trackable(layer, name="layer")
```

Comment:
`Layer` is auto trackable, and has already been added to attr when calling `super`.

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/layers/wrappers.py#L53-L54

Reply:
Sure